### ISSUE TITLE
implement DCGAN with PaddlePaddle

### ISSUE THREAD

**Issue by jacquesqiao** [PaddlePaddle/Paddle#4271]
![image](https://user-images.githubusercontent.com/3048612/30668938-4cda8452-9e11-11e7-9e4f-070748695b14.png)

**Comment by jacquesqiao** [comment]
The GAN support for Paddle is very hacky. It lacks documentation. The GradientMachine is the abstract class, like a neural network. You can use `help()` method in Python to figure out what methods that GradientMachine contains.

**Comment by jacquesqiao** [comment]
The `GradientMachine` calculate gradients for each input layers. Responsible for  the `forward`, `backward` loop call in your network. https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/gradientmachines/GradientMachine.h#L106.  These two virtual function is the backbone of Gradient Machine.
GitHub
PaddlePaddle/Paddle
Paddle - PArallel Distributed Deep LEarning
 

`GAN` will be support smoothly in our near-release new code. But at present, we only have that hack demo for `GAN`.

**Comment by jacquesqiao** [comment]
https://devpost.com/software/dcgan-in-paddlepaddle/

**Comment by jacquesqiao** [comment]
```python
import paddle.v2 as paddle
import pickle as pk
import os
import numpy as np


def get_noise(batch_size, noise_dim=100):
    return numpy.random.normal(size=(batch_size, noise_dim)).astype('float32')

def conv_block_D(ind, ipt, filter_size, num_channels, num_filters, stride, padding, mode=-1):
    act = paddle.activation.Sigmoid() if mode == 1 else paddle.activation.Linear()

    h1 = paddle.layer.img_conv(
        name='conv_D_%d' % ind,
        input=ipt,
        filter_size=filter_size,
        num_channels=num_channels,
        num_filters=num_filters,
        stride=stride,
        padding=padding,
        act=act)

    if mode==0:
        return paddle.layer.prelu(input=h1, partial_sum=1)
    else:
        return h1

def bn_D(ipt):
    return paddle.layer.batch_norm(
        input=ipt,
        act=paddle.activation.SoftRelu())

def conv_block_G(ind, ipt, filter_size, num_channels, num_filters, stride, padding, mode=-1):
    act = paddle.activation.Tanh() if mode==0 else paddle.activation.Relu()

    return paddle.layer.img_conv(
        name='conv_G_%d' % ind,
        input=ipt,
        filter_size=filter_size,
        num_channels=num_channels,
        num_filters=num_filters, stride=stride,
        trans=True,
        padding=padding,
        act=act)

def bn_G(ipt):
    return paddle.layer.batch_norm(
        input=ipt,
        act=paddle.activation.Relu())

def _netG(z, ngf=64, nc=3, func=None):

    nz = z.size
    tconv1 = conv_block_G(1, z, 4, nz, ngf * 8, 1, 0)
    tconv2 = conv_block_G(2, bn_G(tconv1), 4, ngf * 8, ngf * 4, 2, 1)
    tconv3 = conv_block_G(3, bn_G(tconv2), 4, ngf * 4, ngf * 2, 2, 1)
    tconv4 = conv_block_G(4, bn_G(tconv3), 4, ngf * 2, ngf, 2, 1)
    out = conv_block_G(5, bn_G(tconv4), 4, ngf, nc, 2, 1, mode=0)

    return out

def _netD(img, nc=3, ndf=64):

    conv1 = conv_block_D(1, img, 4, nc, ndf, 2, 1, mode=0)
    conv2 = conv_block_D(2, conv1, 4, ndf, ndf*2, 2, 1)
    conv3 = conv_block_D(3, bn_D(conv2), 4, ndf*2, ndf*4, 2, 1)
    conv4 = conv_block_D(4, bn_D(conv3), 4, ndf*4, ndf*8, 2, 1)
    out = conv_block_D(5, bn_D(conv4), 4, ndf*8, 1, 1, 0, mode=1)

    return out # might need to reshape this

def _netD_2(img, nc=3, ndf=64):

    conv1 = conv_block_D(12, img, 4, nc, ndf, 2, 1, mode=0)
    conv2 = conv_block_D(22, conv1, 4, ndf, ndf*2, 2, 1)
    conv3 = conv_block_D(32, bn_D(conv2), 4, ndf*2, ndf*4, 2, 1)
    conv4 = conv_block_D(42, bn_D(conv3), 4, ndf*4, ndf*8, 2, 1)
    out = conv_block_D(52, bn_D(conv4), 4, ndf*8, 1, 1, 0, mode=1)

    return out # might need to reshape this

def get_real(predict_G, pG, noise, batch_indx, fake_size=64):
    with open('/root/.cache/mnist/%d.pkl' % batch_indx) as f:
        r = pk.load(f)
    return np.concatenate((r, get_fake(predict_G, pG, noise, fake_size))),


def get_fake(predict_G, pG, noise, batch_size=128):

    noise = get_noise(batch_size)
    return paddle.layers.infer(predict_G, pG, noise),

# event_handler to print training info
def event_handler_G(event):
    if isinstance(event, paddle.event.EndIteration):
        print "G Pass %d, Batch %d, Cost %f" % (
            event.pass_id, event.batch_id, event.cost)

# event_handler to print training info
def event_handler_D(event):
    if isinstance(event, paddle.event.EndIteration):
        print "D Pass %d, Batch %d, Cost %f" % (
            event.pass_id, event.batch_id, event.cost)


if __name__ == '__main__':

    num_pkl = len(os.listdir('/root/.cache/mnist'))-1
    input_D = paddle.layer.data(name='input_d', type=paddle.data_type.dense_vector(64*64*3))#placeholder
    label_D = paddle.layer.data(name='label_d', type=paddle.data_type.integer_value_sequence(2))#placeholder
    predict_D = _netD(input_D)
    parametersD = paddle.parameters.create(predict_D)
    costD = paddle.layer.cross_entropy_cost(input=predict_D, label=label_D)

    input_G = paddle.layer.data(name='input_g', type=paddle.data_type.dense_vector(100))
    fakes = _netG(input_G)
    predict_G = _netD_2(fakes)
    parametersG = paddle.parameters.create(fakes)
    label_G = paddle.layer.data(name='label_g', type=paddle.data_type.integer_value_sequence(2))#placeholderpaddle.v2.data_type.sparse_binary_vecto
    costG = paddle.layer.cross_entropy_cost(input=predict_G, label=label_G)

    optimizerD = paddle.optimizer.Adam(beta1=0.5, beta2=0.999, epsilon=1e-8)
    optimizerG = paddle.optimizer.Adam(beta1=0.5, beta2=0.999, epsilon=1e-8)

    trainerD = paddle.trainer.SGD(cost=costD,
                                 parameters=parametersD,
                                 update_equation=optimizerD)

    trainerD_2 = paddle.trainer.SGD(cost=costD,
                                 parameters=parametersD,
                                 update_equation=optimizerD)

    trainerG = paddle.trainer.SGD(cost=costG,
                                 parameters=parametersG,
                                 update_equation=optimizerG)


    for Epoch in range(100):
        for batch_indx in range(1,num_pkl+1):

            d_params = trainerD.parameters
            for k in d_params.keys():
                trainerD_2.set(k+'2', d_params.get(k))

            data_R = get_real(predict_D, trainerD.parameters, data, batch_indx)
            pred_D = paddle.infer(predict_D, trainerD.parameters, data_R)
            data_F = get_fake(predict_G, trainerG.parameters, noise)
            pred_G = paddle.infer(predict_G, trainerG.parameters, data_F)

            c_D = pred_D == ground_D
            score_D = c_D.count_nonzero()

            c_G = pred_G == ground_G
            score_G = c_G.count_nonzero()

            if score_D < score_G:
                trainerD.train(
                    reader=paddle.batch(
                        paddle.reader.shuffle(
                            data_R, buf_size=8192),
                        batch_size=128),
                    event_handler=event_handler_D,
                    num_passes=3)
            else:

                trainerG.train(
                    reader=paddle.batch(
                        paddle.reader.shuffle(
                            data_F, buf_size=8192),
                        batch_size=128),
                    event_handler=event_handler_G,
                    num_passes=3)
```
